Subscriptions

dperpeet opened the pull request:
left undone:
* potentially: auto attach on registered system with no valid subscription

Change via web interface doesn't propagate to cockpit after yum update

https://github.com/cockpit-project/cockpit/wiki/Feature:-Subscriptions

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,124 @@
+        "latest/cockpit",
+        "latest/mustache",
+        "subscriptions/subscriptions"
+        ], function($, cockpit, Mustache, subscriptions) {
```
You don't need to require all of these, if you're only going to use ```subscriptions```.

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,276 @@
+
+    var subscriptions = { };
+
+    PageSubscriptionsRegister.prototype = {
```
While it's not a big deal, you don't actually need to create class here at all. Just run code in the various event handlers like ```dlg.on('shown.bs.modal', function() { /* here */});````

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,276 @@
+            var process = cockpit.spawn(args, { directory: '/', superuser: true })
+                .done(function(output) {
+                    // parse for success
+                    var success = (output.indexOf('system has been registered') > -1);
```
If you're going to be parsing output, you need to make sure to spawn with the ```LC_ALL=C``` environment variable to disable any localization.

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,276 @@
+         * Parses output like:
+         *
+         * bios_vendor:LENOVO
+         * bios_version:8CET46WW
```
Does it?

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,276 @@
+        function parse_multiple_subscriptions(text) {
+            var ret = [ ];
+            var parts = text.split('\n\n');
+            for (var i=0;i<parts.length;++i) {
```
Code style: Suggest using spaces around operators.

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,276 @@
+
+        function get_subscriptions_info() {
+            cockpit.spawn(['subscription-manager', 'list'],
+                    { directory: '/', superuser: true })
```
Ditto about ```LC_ALL=C```

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,276 @@
+                })
+                .fail(function(ex) {
+                    is_updating.hide();
+                    if (ex.problem != 'not-authorized')
```
Out of interest, how are you handling lack of privileges to perform the tasks?

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,276 @@
+                 * since older versions of subscription-manager don't deliver this via DBus
+                 * note: subscription-manager needs superuser privileges
+                 */
+                get_subscriptions_info();
```
You need to do some sort of rate limiting, or you could easily get into a situation where you are invoking ```subscription-manager list``` 10 times all at once.

stefwalter commented:
The approach looks good. Didn't try this yet, but posted some initial feedback.

dperpeet commented:
Currently missing: org input on registration, test for interaction with yum

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+        require([
+        "subscriptions/subscriptions"
+        ], function(subscriptions) {
+            //
```
Including an AMD package should not have side-effects like this. Maybe add an subscriptions.init() function and call it here?

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,420 @@
+
+/* global jQuery   */
+/* global cockpit  */
+/* global po */
```
None of these should be necessary.

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,420 @@
+], function($, cockpit, po, Mustache) {
+    "use strict";
+
+    cockpit.locale(po);
```
You probably also want a call to ```cockpit.translate()``` to translate the page.

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+    <div id="subscriptions-register-failure" class="container-fluid" hidden>
+      <div class="alert alert-danger" role="alert">
+        <span class="pficon-layered > pficon pficon-error-octagon + pficon pficon-error-exclamation"></span>
+          <span><b>System couldn't be registered.</b></span>
```
All translatable strings should have ```translatable="yes"``` as an attribute.

stefwalter commented:
![screenshot from 2015-03-02 15 42 32](https://cloud.githubusercontent.com/assets/795070/6442788/c424c68e-c0f2-11e4-934f-8b62cd2773e4.png)

The font size of the status text doesn't match that in the design.

stefwalter commented:
Removing subscriptions results in this display:

![screenshot from 2015-03-02 15 43 58](https://cloud.githubusercontent.com/assets/795070/6442813/f4f6f994-c0f2-11e4-9c86-ef17ce057b85.png)

This doesn't match the wireframes.

stefwalter commented:
Spacing above the ```[Register System]``` button needs to be fixed.

stefwalter commented:
The [x] in the alert is not centered.

stefwalter commented:
![screenshot from 2015-03-02 15 47 03](https://cloud.githubusercontent.com/assets/795070/6442877/6362144a-c0f3-11e4-8406-d2080bc322ae.png)

 * The field labels need to use ```<label class="control-label" for="xxxx">URL</label>```
   * You can see this here: https://www.patternfly.org/wp-content/uploads/patternfly/tests/form.html
   * This allows you to also click on labels and the control will be focused.
 * Needs either a prefilled URL field, or a redesign if you need to leave this blank
    * Probably needs a checkbox in the latter case. Would need to check with @andreasn 
 * Needs placeholder text in the User name field
 * Pressing <enter> in the Password field has no effect

stefwalter commented:
![screenshot from 2015-03-02 15 52 35](https://cloud.githubusercontent.com/assets/795070/6442985/275a7f0e-c0f4-11e4-8b03-fe12bcf72b1c.png)

 * During registration, the [Register System] button shouldn't be visible, or shouldn't be clickable.
 * @andreasn should the registration happen while the dialog is open, or should it happen in the page itself?

stefwalter commented:
![screenshot from 2015-03-02 15 54 35](https://cloud.githubusercontent.com/assets/795070/6443026/6fc14bb0-c0f4-11e4-8b51-cbb61cd7342b.png)

 * Multiple messages are shown at the same time.

stefwalter commented:
Needs integration tests

stefwalter commented:
Minus the above issues. This works really well. Nice work :)

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+        <div class="panel-body">
+          <table class="cockpit-form-table">
+            <tbody>
+              <tr><td>System identity</td><td><span id="subscription-system-identity"></span></td></tr>
```
What are these cells used for?

dperpeet replied:
> What are these cells used for?

@andreasn We can get additional system info (such as registration ID) from subscription manager. I thought it might be useful, but it's not crucial. What do you think?

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,420 @@
+            if (info.length > 0) {
+                subscription_info.show();
+                /* render and add the new ones */
+                for (var i=0;i<info.length;++i)
```
Code style: Spaces around operators

stefwalter commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,420 @@
+            show_progress_message('Retrieving current subscription details.');
+
+            cockpit.spawn(['subscription-manager', 'list'],
+                    { directory: '/', superuser: true, environ: ['LC_ALL=C'] })
```
Suggest adding a ```TODO:``` explaining that this will result in non-translated values, and that we'll need to wait until a DBus API in order to rectify this.

stefwalter commented:
The screen remains blank for about 1 second before showing the "Retrieving current subscription details." status message.

stefwalter commented:
You need to install using a ```Makefile.am``` like in the ```pkg/docker``` package. In addition a subpackage in ```tools/cockpit.spec```. I can help you make a subpackage if you want.

stefwalter commented:
```
--- a/pkg/subscriptions/subscriptions.html
+++ b/pkg/subscriptions/subscriptions.html
@@ -9,7 +9,10 @@
     <!-- different containers, depending on current subscription status: updating, unregistered, registered -->
     <div id="subscriptions-register-failure" class="container-fluid" hidden>
       <div class="alert alert-danger" role="alert">
-        <span class="pficon-layered > pficon pficon-error-octagon + pficon pficon-error-exclamation"></span>
+        <span class="pficon-layered">
+          <span class="pficon pficon-error-octagon"></span>
+          <span class="pficon pficon-error-exclamation"></span>
+        </span>
           <span><b>System couldn't be registered.</b></span>
           <span id="subscriptions-register-failure-reason"></span>
         <button type="button" class="close" aria-label="Close"
@@ -30,7 +33,10 @@
     </div>
     <div id="subscriptions-status-subscribe-failure" class="container-fluid" hidden>
       <div class="alert alert-danger" role="alert">
-        <span class="pficon-layered > pficon pficon-error-octagon + pficon pficon-error-exclamation"></span>
+        <span class="pficon-layered">
+          <span class="pficon pficon-error-octagon"></span>
+          <span class="pficon pficon-error-exclamation"></span>
+        </span>
           <span><b>System isn't subscribed.</b></span>
           <span id="subscriptions-status-subscribe-failure-reason">You cannot receive software updates.</span>
         <button type="button" class="close" aria-label="Close"
```

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+    <div id="subscriptions-status-subscribe-success" class="container-fluid" hidden>
+      <div class="alert alert-success" role="alert">
+        <span class="pficon pficon-ok"></span>
+        <span><b>System successfully subscribed</b> You will now receive software updates.</span>
```
Needs punctuation at the end of the sentence.

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+      <span style="display: block; text-align: center;">
+        In order for your system to receive updates, your system must be registered.
+      </span>
+      <button class="btn btn-default btn-primary" id="subscriptions-register" style="display: block; margin: 0px auto;">
```
We shouldn't need inline styles like this.

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+      </button>
+    </div>
+    <div id="subscriptions-updating" class="container-fluid" hidden>
+      <span id="subscriptions-update-message" style="padding-left: 10px;"></span>
```
Ditto about inline styles.

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+    </div>
+    <div id="subscriptions-updating" class="container-fluid" hidden>
+      <span id="subscriptions-update-message" style="padding-left: 10px;"></span>
+      <div class="spinner spinner-sm" style="float: left;"></div>
```
You can use the ```pull-left``` class instead of the inline style here.

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+          <span><b>System couldn't be registered.</b></span>
+          <span id="subscriptions-register-failure-reason"></span>
+        <button type="button" class="close" aria-label="Close"
+                onclick="$('#subscriptions-register-failure').hide();">
```
Bootstrap alerts can be closed without onclick handlers: http://getbootstrap.com/components/#alerts-dismissible

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+        <span class="pficon pficon-ok"></span>
+        <span><b>System successfully subscribed</b> You will now receive software updates.</span>
+        <button type="button" class="close" aria-label="Close"
+                onclick="$('#subscriptions-status-subscribe-success').hide();">
```
http://getbootstrap.com/components/#alerts-dismissible

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+          <span><b>System isn't subscribed.</b></span>
+          <span id="subscriptions-status-subscribe-failure-reason">You cannot receive software updates.</span>
+        <button type="button" class="close" aria-label="Close"
+                onclick="$('#subscriptions-status-subscribe-failure').hide();">
```
http://getbootstrap.com/components/#alerts-dismissible

stefwalter commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,157 @@
+      </div>
+    </div>
+    <div id="subscriptions-unregistered" class="container-fluid" hidden>
+      <span style="display: block; text-align: center;">
```
Suggest adding a semantic class and a definition in a ```<style>``` block above instead of an inline style.

stefwalter commented:
I see tons of messages like this in the journal:

```
cockpit-bridge: Invalid username or password. To create a login, please visit https://www.redhat.com/wapps/ugc/register.html16:07
cockpit-bridge: Password: 16:05
cockpit-bridge: Warning: Password input may be echoed.16:05
cockpit-bridge: passwd = fallback_getpass(prompt, stream)16:05
cockpit-bridge: /usr/lib64/python2.7/getpass.py:83: GetPassWarning: Can not control echo on the terminal.16:05
cockpit-bridge: Invalid username or password. To create a login, please visit https://www.redhat.com/wapps/ugc/register.html
```

Likely you need to run your commands with ```error: "output"``` or we need a new ```error: null``` option to hide this stuff?

andreasn commented:
> should the registration happen while the dialog is open, or should it happen in the page itself?

@stefwalter I realized I hadn't specified this when I was out walking the other day. Sorry for missing to specify that in the design. I think it should happen in the dialog itself, since if you, say mistyped the password and the system happily tries to log you in, but fail due to this, you still have the opportunity to fix it without having to reopen the dialog.

The spinner should go on the left to the buttons, aligned to the left edge of the dialog.

dperpeet commented:
> Needs either a prefilled URL field, or a redesign if you need to leave this blank
> 
>     Probably needs a checkbox in the latter case. Would need to check with @andreasn

as Discussed with @andreasn I think we should go for having a dropdown with two choices: 'Default' (let subscription manager choose) and 'Custom URL', which opens a text input field for the user to fill in a url.

dperpeet commented:
> Why is there a heading called 'System'?

It can happen that the system is registered, but not attached. In that case, our call to subscription manager with "register --auto-attach" will fail, but it's not readily apparent to the user why.
@andreasn @stefwalter Decision time: Add option to attach subscriptions (auto-attach) to a registered system? This is a quick fix and might enable users to fix some configurations. Of course from there it's easy to add buttons to unregister or remove subscriptions...

andreasn commented:
What would those configurations be?
A system that expired, because the days ran out?

stefwalter commented:
>  @andreasn @stefwalter Decision time: Add option to attach subscriptions (auto-attach) to a registered system? This is a quick fix and might enable users to fix some configurations. Of course from there it's easy to add buttons to unregister or remove subscriptions...

I think this should be done with just the basics for now. It's easy to add these features later.

dperpeet commented:
> Needs integration tests

@stefwalter After some consideration, I think the most thorough way to test this would be to implement a mock subscription-manager for successful registration, the real one for everything else. What do you think?

stefwalter commented:
> After some consideration, I think the most thorough way to test this would be to implement a mock subscription-manager for successful registration, the real one for everything else. What do you think?

Traditionally an integration test means that this is tested against the "real thing". Obviously there are various shades of gray here. But have you taken a look at setting up a CandlePin instance that the real subscription-manager can talk to?

http://www.candlepinproject.org/docs/candlepin/setup.html

stefwalter commented:
The candle pin instance would be deployed as an alternate image, similar to how we have an ipa image in our tests.

andreasn commented:
Having issues with the css paths, possibly as an outcome of @stefw landing #1869, and the javascript ends up as a 404

stefwalter commented:
Could you post the failed URL?

stefwalter commented:
I hit the wrong button .... oops.

andreasn commented:
https://192.168.122.136:9090/cockpit/$e3b6a627e036f94f98640708a31e9fc8572ec128/base/cockpit.js

stefwalter commented:
If @dperpeet hasn't rebased onto master, then you'll need to follow the instructions on cockpit-devel (including ```sudo rm...```) each time you change to a branch on either side of the changes that landed today.

dperpeet commented:
@andreasn rebased

andreasn commented:
Finally got this working.
Very nice stuff already!

![screenshot from 2015-03-05 10 25 32](https://cloud.githubusercontent.com/assets/209895/6505551/183a82d2-c322-11e4-9213-b68a355fd109.png)

The URL button doesn't fill the width like our other dialog dropdowns, and the url field is probably better on it's own line. The URLs can potentially be very long, so good to give it plenty of space.

dperpeet commented:
> The URL button doesn't fill the width like our other dialog dropdowns, and the url field is probably better on it's own line. The URLs can potentially be very long, so good to give it plenty of space.

I agree about the field not being wide enough.

But I don't think it's good to have the url on it's own line. We only have three input lines right now. Having an optional one will make everything move around when it's hidden/shown and the field description "URL" on the left will move also. When trying this option, it seemed less clear than one line.

andreasn commented:
I currently run into this, due to the dialog still being around somehow (?)

![screenshot from 2015-03-05 10 49 01](https://cloud.githubusercontent.com/assets/209895/6505917/86246d64-c325-11e4-9209-3b267cf54264.png)

stefwalter commented:
> I currently run into this, due to the dialog still being around somehow (?)

Guess: Looks like you're just hiding the dialog. You need to use the bootstrap modal functions to get rid of it http://getbootstrap.com/javascript/#modals-methods

dperpeet commented:
> Guess: Looks like you're just hiding the dialog. You need to use the bootstrap modal functions to get rid of it http://getbootstrap.com/javascript/#modals-methods

yes, thanks. I already fixed it. It happened when moved registration progress into the dialog...

dperpeet commented:
> The URL button doesn't fill the width like our other dialog dropdowns, and the url field is probably better on it's own line. The URLs can potentially be very long, so good to give it plenty of space.

@andreasn I think the url on a separate line can work:
![screenshot from 2015-03-05 15 35 52](https://cloud.githubusercontent.com/assets/8711649/6506740/7af43abe-c34d-11e4-82e5-665e3450a6f4.png)

andreasn commented:
Error state mockups for the dialog

![Mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/subscriptions/subscriptions-errors.png)

andreasn commented:
if you add class="top" to the td containing the string URL it will align nicely to the top

dperpeet commented:
> Error state mockups for the dialog

I will adress this in a separate issue to reduce scope. This is because all subscription-manager tells us is that registration failed, the rest is text. We can parse this for known errors (wrong url, wrong login/password combination) and display those in the manner you designed. Our fallback option would be a generic error.

dperpeet commented:
Updated error display:
![screenshot from 2015-03-06 11 54 49](https://cloud.githubusercontent.com/assets/8711649/6524495/70454c86-c3fa-11e4-8cff-387ebeefc4f5.png)
![screenshot from 2015-03-06 11 55 09](https://cloud.githubusercontent.com/assets/8711649/6524497/7273f656-c3fa-11e4-9bb5-2b29fcdd5b03.png)
Fallback for generic error, displays output directly:
![screenshot from 2015-03-06 11 57 15](https://cloud.githubusercontent.com/assets/8711649/6524498/77633eb0-c3fa-11e4-910b-4321228fac98.png)

stefwalter commented:
Two commits for this branch here, the first one you can squash into another commit:

https://github.com/stefwalter/cockpit/commits/subscriptions

dperpeet commented:
I made the decision that errors shown in the register dialog will persist through closing/reopening the dialog.

@andreasn Should I remove the errors when the dialog is shown?

stefwalter commented:
> Should I remove the errors when the dialog is shown?

I think so, otherwise the user sees errors before they've even started their task. This is a common design flaw.

andreasn commented:
> Should I remove the errors when the dialog is shown?

Yes. Opening it again should be as a fresh try as possible.

andreasn commented:
So I only tested this on a Fedora system, will test more on a system that requires a proper subscription and hopefully we can soon have a little SAM server for testing.
The dialog works great now!

On the subscription page itself, it says "System" as a header. I wonder if "Subscription" or "Subscription status" would be a more describing label for that, but not a biggie. Apart from that, all good!

dperpeet commented:
> On the subscription page itself, it says "System" as a header. I wonder if "Subscription" or "Subscription status" would be a more describing label for that, but not a biggie. Apart from that, all good!

I changed it to "Installed Products" to be more consistent with what the user sees if there are subscriptions: "Installed Product (x)"

andreasn commented:
> I changed it to "Installed Products" to be more consistent with what the user sees if there are subscriptions: "Installed Product (x)"

Good call!

dperpeet commented:
thanks for the feedback @andreasn, I fixed that duplicate entry bug

petervo commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,213 @@
+      </span>
+      <br/>
+      <button class="btn btn-default btn-primary" id="subscriptions-register">
+        Register system
```
Translate?

petervo commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,213 @@
+          <span class="pficon pficon-error-octagon"></span>
+          <span class="pficon pficon-error-exclamation"></span>
+        </span>
+        <span class="subscriptions-emphasize" translatable="yes">
```
Not sure translatable is needed here

petervo commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,213 @@
+            <tbody>
+              <tr>
+                <td></td>
+                <td><span>No subscriptions on the system.</span></td>
```
Translate?

petervo commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,213 @@
+                           value="subscription.rhn.redhat.com">
+                    <div id="subscriptions-register-url-note" hidden>
+                      <span class="help-block" id="subscriptions-register-url-message">
+                        Unable to reach the server.
```
Translate?

petervo commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,213 @@
+                <td>
+                  <input id="subscription-register-password" class="form-control" type="password">
+                  <div id="subscriptions-register-password-note" hidden>
+                    <span class="help-block">
```
Translate?

petervo commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,213 @@
+              </tr>
+            </table>
+            <div class="has-error" hidden>
+              <span id="subscriptions-register-general-error" class="help-block">
```
Translate?

petervo commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,213 @@
+          <div class="modal-footer">
+            <div id="subscriptions-registering" class="pull-left" hidden>
+              <div class="spinner spinner-sm pull-left"></div>
+              <span>Registering system.</span>
```
Translate?

petervo commented on pkg/subscriptions/subscriptions.html:
```diff
@@ -0,0 +1,213 @@
+
+    <script id="subscription-template" type="x-template/mustache">
+      <div class="panel panel-default">
+        <div class="panel-heading">
```
Translate?

petervo commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,502 @@
+        if ($('#subscription-register-url').val() === 'Custom URL')
+          url = $('#subscription-register-url-custom').val().trim();
+        var username = $('#subscription-register-username').val().trim();
+        var password = $('#subscription-register-password').val().trim();
```
Not sure we want to trip passwords? Are you sure that passwords ending in a space are illegal?

dperpeet replied:
> Not sure we want to trim passwords? Are you sure that passwords ending in a space are illegal?

You're absolutely right. It always annoys me when ui's make that kind of choice for me.

petervo commented on pkg/subscriptions/subscriptions.js:
```diff
@@ -0,0 +1,502 @@
+        function register_system(url, username, password) {
+            if (action_in_progress()) {
+                $('#subscriptions-register-failure-reason').text(
+                  'Unable to register at this time because a call to subscription manager ' +
```
Translate?

dperpeet replied:
`subscriptions-register-failure` doesn't exist anymore, so I changed this into a console log message. This shouldn't happen on a ui-path anyway: the "Register" button is hidden during status check and while performing registration.

petervo commented:
Hitting register with no pw, results in an ugly looking warning.

/usr/lib64/python2.7/getpass.py:83: GetPassWarning: Can not control echo on the terminal. passwd = fallback_getpass(prompt, stream) Warning: Password input may be echoed. Password:

Not really our bug, but still probably shouldn't show that, maybe a little FE validation? WDYT?

dperpeet commented:
> Hitting register with no pw, results in an ugly looking warning.

This bug crept in when I made the url optional... subscription-manager will prompt for username and/or pass from stdin if they aren't passed. Therefore I will always pass them to the cli, even if empty.

dperpeet commented:
Added message if there is a problem contacting subscription-manager via dbus or cli:
![screenshot from 2015-03-10 17 00 59](https://cloud.githubusercontent.com/assets/8711649/6579086/58cf3458-c748-11e4-933d-dd9673fd19b6.png)

dperpeet commented:
updated: username and password now cannot be empty, since subscription-manager cli prompts in shell if they are.
![screenshot from 2015-03-11 20 45 34](https://cloud.githubusercontent.com/assets/8711649/6605381/c8960860-c82f-11e4-95bf-32dbb56a8588.png)

andreasn commented:
I haven't tried it in action yet, but based on the screenshots, this seems good!
Maybe: "Login field cannot be empty" is clearer and same for the password, but no biggie.

dperpeet commented:
> Maybe: "Login field cannot be empty" is clearer and same for the password, but no biggie

I'm not sure that would be much clearer. Technically the fields can be empty, but the process will fail if the actual username or password are empty... I'll defer to your decision on this. Do you think it should be changed?

andreasn commented:
No, lets keep it like it is now.

andreasn commented:
Tested this on both a RHEL7 and a Fedora 21 system and everything seems to work really well now!

petervo commented:
Looks good, last thing it i try to access as user that doesn't have permissions I get the "please install subscription-manager" message. Are you able to tell the difference between the command missing and permission denied? error codes might work.

dperpeet commented:
Differentiated message. Not found:
![screenshot from 2015-03-12 16 24 40](https://cloud.githubusercontent.com/assets/8711649/6621175/53e0732e-c8d5-11e4-8b78-089a33acdabe.png)

Not authorized:
![screenshot from 2015-03-12 16 25 02](https://cloud.githubusercontent.com/assets/8711649/6621179/57bd8edc-c8d5-11e4-8763-724bf8ae3be9.png)

dperpeet commented:
I also added the subscription-manager dependency back in, per @stefwalter's suggestion

andreasn commented:
> Differentiated message

Looks good.

mvollmer commented:
```
logger: unknown priority name: 3
CalledProcessError: Command 'logger -p3 --tag check-journal BEFORE BOOT' returned non-zero exit status 1
```

Wot?

dperpeet commented:
> Strange as this looks, this means we can't install those images before investigating and fixing this.

The error was `logger: unknown priority name: 3`
I changed this from `-p3` to `-p user.err`, which should make a little clearer what was intended here and conforms to what logger expects to see. In the man pages I see no example with no space between `-p` and the number.
I also added a comment that we insert an error because our journal view always shows errors by default. This way the test doesn't have to change the view to detect the added lines.

dperpeet commented:
> I changed this from -p3 to -p user.err

Even if the failure was random, the proposed change should make the test a bit clearer in any case.
